### ISSUE TITLE
Fix strict aliasing warnings and re-enable the optimization

### ISSUE THREAD

**Issue by turol** [assimp/assimp#2147]
#2143 disabled strict aliasing optimization because our code is broken. We want to re-enable it and that means fixing all the warnings. Use GCC (maybe Clang works too) with options like this:
`cmake -DCMAKE_CXX_FLAGS="-O2 -fstrict-aliasing -Wstrict-aliasing=3" `
And fix all the warnings it gives. Note that the warning level is "backwards", level 1 finds more issues than level 2 which finds more issues than level 3. Start with 3 and work your way backwards. You'll have to deal with nasty code like #1994. Once all level 1 warnings are fixed we can remove `-fno-strict-aliasing` from CMakeLists.txt

**Comment by kimkulling** [comment]
Closed as far as I know.

**Comment by turol** [comment]
Not fixed. You'll know it's fixed when `-fno-strict-aliasing` has been removed from `CMakeLists.txt`.

**Comment by kimkulling** [comment]
Ops, right. Thanks a lot for the hint!

**Comment by kimkulling** [comment]
Will put this to 5.1 release.

**Comment by Optimizer0** [comment]
@turol Sir I fixed all level 3 warnings. (they were mainly in m3d) I used unions to fix the type punning.

But to fix level 2 warnings I will have to modify core structures of Assimp like aiNode. Is this fine?

**Comment by turol** [comment]
Using `union` is not necessarily enough, you might have to use explicit `memcpy`. I can't remember the specific rules when it's safe and when not'

`aiNode` class is part of the ABI and changing it requires bumping the major version. Ask @kimkulling when is a good time to do that.

**Comment by Optimizer0** [comment]
Err I definitely am not an expert but I believe unions are enough for the cases I did so far. they were something like this. 

`
float f;

f = some_value_based_on_a_switch;
h->cmd[i].arg[k] = *((uint32_t*)&f);
`

to

`
union {
float f;
uint32_t f_uint32;
};

f = some_value_based_on_a_switch;
h->cmd[i].arg[k] = f_uint32;
`

I ran some tests and all worked BUT I will continue to test them.

About changing core structures, changing those structures are not essential. There definitely are ways to fix the warnings without touching them but I think I'll be able to optimize better If I'm allowed to change them. (I used 'I think' because I need study the code more before saying anything for sure)

**Comment by turol** [comment]
Try these links:
https://stackoverflow.com/questions/54762186/unions-aliasing-and-type-punning-in-practice-what-works-and-what-does-not
https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8

So it's valid in C but no in C++.

This
```
u.f = 1.0f;
u.f_uint32_t = 0;
printf("%f\n", u.f);
```

Could print either 0.0 or 1.0 depending on the compiler and optimization settings.

**Comment by Optimizer0** [comment]
I didn't know this. Thank you for pointing out. The two links you gave were a lot of helpful. I did a bit of reading and saw what you said is indeed true. (this was also helpful : https://stackoverflow.com/questions/11373203/accessing-inactive-union-member-and-undefined-behavior) 

I will better study aliasing before continuing to fix the warnings. I will also post any changes I make to the code here to get your approval first, if that is fine by you.

**Comment by tellypresence** [comment]
@turol Looks like `-fno-strict-aliasing` only remains in place for `mips64` architecture in `CMakeLists.txt`
https://github.com/assimp/assimp/blob/ac5988422a7f1898b25c13fb5fb068c646a4c544/CMakeLists.txt#L319-L321

Can this be closed?

**Comment by turol** [comment]
No, it's still there in the else branch 5 lines below. Also later in clang and mingw sections. Don't close this issue until they're all gone and the library is tested to still work.

**Comment by tellypresence** [comment]
Feature request issues tracked via epic #6183
